File: iosApp/iosApp/Bridge/AlbumComponent.swift
import HotwireNative
import sharedKit

class AlbumComponent: BridgeComponent {
    override class var name: String { "album" }

    private var viewModel: WebViewModel? {
        let viewController = delegate?.destination as? WebViewController
        return viewController?.viewModel
    }

    override func onReceive(message: Message) {
        switch message.event {
        case "play":
            handlePlayEvent(message)
        case "playBeginWith":
            handlePlayBeginWithEvent(message)
        default:
            break
        }
    }

    private func handlePlayEvent(_ message: Message) {
        guard let data: AlbumData = message.data() else { return }
        viewModel?.playAlbum(albumId: data.albumId)
    }

    private func handlePlayBeginWithEvent(_ message: Message) {
        guard let data: AlbumData = message.data(), let songId = data.songId else { return }
        viewModel?.playAlbumBeginWith(albumId: data.albumId, songId: songId)
    }
}

extension AlbumComponent {
    struct AlbumData: Decodable {
        let albumId: Int64
        let songId: Int64?
    }
}

File: androidApp/src/main/java/org/blackcandy/android/bridge/AlbumComponent.kt
package org.blackcandy.android.bridge

import dev.hotwire.core.bridge.BridgeComponent
import dev.hotwire.core.bridge.BridgeDelegate
import dev.hotwire.core.bridge.Message
import dev.hotwire.navigation.destinations.HotwireDestination
import kotlinx.serialization.Serializable
import org.blackcandy.android.fragments.web.WebFragment
import org.blackcandy.shared.viewmodels.WebViewModel

class AlbumComponent(
    name: String,
    private val delegate: BridgeDelegate<HotwireDestination>,
) : BridgeComponent<HotwireDestination>(name, delegate) {
    private val viewModel: WebViewModel
        get() {
            val fragment = delegate.destination.fragment as WebFragment
            return fragment.viewModel
        }

    override fun onReceive(message: Message) {
        when (message.event) {
            "play" -> handlePlayEvent(message)
            "playBeginWith" -> handlePlayBeginWithEvent(message)
        }
    }

    private fun handlePlayEvent(message: Message) {
        val data = message.data<AlbumData>() ?: return
        viewModel.playAlbum(data.albumId)
    }

    private fun handlePlayBeginWithEvent(message: Message) {
        val data = message.data<AlbumData>() ?: return
        viewModel.playAlbumBeginWith(data.albumId, data.songId!!)
    }

    @Serializable
    data class AlbumData(
        val albumId: Long,
        val songId: Long?,
    )
}
